pyteal/ast/router.py

```diff
@@ -274,13 +280,194 @@ class CondNode:
 CondNode.__module__ = "pyteal"
 
 
+@dataclass(frozen=True)
+class CondWithMethod:
+    """CondWithMethod unifies the generation to CondNode from compilation w/wo frame pointer.
+
+    It takes following arguments:
+    - method_sig: str
+    - condition: Expr | int, which matches the on_completion conditions
+    - method: ABIReturnSubroutine, assuming that the method is ABI routable
+    """
+
+    method_sig: str
+    condition: Expr | int
+    method: ABIReturnSubroutine
+
+    def to_cond_node(self, use_frame_pt: bool = False) -> CondNode:
+        walk_in_cond = Txn.application_args[0] == MethodSignature(self.method_sig)
+
+        if not (isinstance(self.condition, Expr) or self.condition == 1):
+            raise TealInputError("Invalid condition input for CondWithMethod")
+
+        res = ASTBuilder.wrap_handler(True, self.method, use_frame_pt=use_frame_pt)
+        if isinstance(self.condition, Expr):
+            res = Seq(Assert(self.condition), res)
+        return CondNode(walk_in_cond, res)
+
+
+CondWithMethod.__module__ = "pyteal"
+
+
 @dataclass
 class ASTBuilder:
     conditions_n_branches: list[CondNode] = field(default_factory=list)
+    methods_with_conds: list[CondWithMethod] = field(default_factory=list)
+
+    @staticmethod
+    def __handler_validity_check(
```

Comment:
It was a bit confusing to me to read this method name and try to understand why it would return a value. So maybe a renaming to something like `_filter_invalid_and_typecast()` would be a bit clearer? (essentially this is a type cast, even though now `mypy` is smart enough to figure this out without you explicitly using `cast()`)

Reply:
sure, I think this method serves following purpose: filter out unroutable ABIReturnSubroutine + Expr + normal subroutine, leave only routable ABIReturnSubroutine passing. So your naming makes the most sense, probably I will take `__filter_invalid_handlers_and_typecast`.